feat: Add collision completed listener

christian-mindfulness opened the pull request:
<!-- Exclude from commit message -->
<!--
The title of your PR on the line above should start with a [Conventional Commit] prefix
(`fix:`, `feat:`, `docs:`, `test:`, `chore:`, `refactor:`, `perf:`, `build:`, `ci:`,
`style:`, `revert:`). This title will later become an entry in the [CHANGELOG], so please
make sure that it summarizes the PR adequately.

Don't remove the "exclude from commit message" comments below. They are used to prevent
the PR description template from being included in the git log.
Only change the "Replace this text" parts.
-->

# Description
<!--
Provide a description of what this PR is doing.
If you're modifying existing behavior, describe the existing behavior, how this PR is changing it,
and what motivated the change. If this is a breaking change, specify explicitly which APIs were
changed.
-->
<!-- End of exclude from commit message -->
Add a new variable to the `CollisionDetection` class.  Call `notifyListeners` at the end of the collision detection step, so that users can add a listener to their code to know when this step has finished.  The variable is a basic implementation of a `Listenable` class, since it needs no more complexity.

<!-- Exclude from commit message -->
## Checklist
<!--
Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes with `[x]`. If some checkbox is not applicable, mark it as `[-]`.
-->

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.


## Breaking Change?
<!--
Would your PR require Flame users to update their apps following your change?

If yes, then the title of the PR should include "!" (for example, `feat!:`, `fix!:`). See
[Conventional Commit] for details. Also, for a breaking PR uncomment and fill in the "Migration
instructions" section below.
-->

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

<!--
### Migration instructions

If the PR is breaking, uncomment this header and add instructions for how to migrate from the
currently released version to the new proposed way.
-->

## Related Issues
<!--
Indicate which issues this PR resolves, if any. For example:

Closes #1234
!-->
<!-- End of exclude from commit message -->
Closes #2849 

<!-- Exclude from commit message -->
<!-- Links -->
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md
<!-- End of exclude from commit message -->

christian-mindfulness commented on packages/flame/lib/src/collisions/collision_detection.dart:
```diff
@@ -13,6 +14,7 @@ abstract class CollisionDetection<T extends Hitbox<T>,
 
   List<T> get items => broadphase.items;
   final _lastPotentials = <CollisionProspect<T>>[];
+  final collisionsCompleted = CollisionDetectionCompletionNotifier();
```
Name changed in latest commit https://github.com/flame-engine/flame/pull/2896/commits/834c950297ab5b0659eddc16dc24c4dc3c7b2da2

spydon commented on packages/flame/lib/src/collisions/collision_detection.dart:
```diff
@@ -161,3 +166,24 @@ abstract class CollisionDetection<T extends Hitbox<T>,
 }
+
+/// Add a basic class which implements a Listener.
+class CollisionDetectionCompletionNotifier implements Listenable {
```
Ah, now I remember. :man_facepalming: 
I think it is probably better to go with `with ChangeNotifier` anyways, so that we don't have to do the re-implementation, and then we can just put an ignore lint lint above the overridden method.

christian-mindfulness commented:
> Some comments, and missing tests

I appreciate the need for tests, but the testing framework is very confusing to me.  @spydon would you be able to suggest a change for this?

christian-mindfulness commented:
> ...
> and then just adding that in the end of this file: https://github.com/flame-engine/flame/blob/main/packages/flame/test/collisions/collision_callback_test.dart

Thanks.  Done that and managed to get the test to run, but fails:
  Expected: <10>
    Actual: <0>
I *think* the code might be adding the listeners, but not running collision detection and so the counter is not incremented?

christian-mindfulness commented:
> @christian-mindfulness did you add the `game.update(0);`? I forgot to add that at first.

Thanks the one, thank you.

spydon commented on doc/flame/collision_detection.md:
```diff
@@ -120,6 +120,41 @@ other way around. If there are no intersections with the edges on a solid hitbox
 
+### Collision order
+
+If a `PositionComponent` collides with more than one other object within a given time step, then
```
```suggestion
If a `Hitbox` collides with more than one other `Hitbox` within a given time step, then
```

spydon commented on doc/flame/collision_detection.md:
```diff
@@ -120,6 +120,41 @@ other way around. If there are no intersections with the edges on a solid hitbox
+### Collision order
+
+If a `PositionComponent` collides with more than one other object within a given time step, then
+the `onCollision` callbacks will be called in an essentially random order.  In some cases this can
+be a problem, such as in a bouncing ball game where the trajectory of the ball can differ depending
+on which other object was hit first.  To help resolve this the `collisionsCompletedNotifier`
+listener can be used - this triggers at the end of the collision detection process.
+
+An example of how this might be used is to add a local variable in your `PositionComponent` to save
+the other components with which this one is colliding:
+`List<PositionComponent> collisionComponents = [];`.  The `onCollision` callback is then used to
+save all the other `PositionComponent`s to this list:
```
All of these sentences start with two spaces, there should only be one to be consistent with the rest of the docs.

christian-mindfulness replied:
Sorry, that's a British English thing, I've changed that.

spydon commented on doc/flame/collision_detection.md:
```diff
@@ -120,6 +120,41 @@ other way around. If there are no intersections with the edges on a solid hitbox
+listener can be used - this triggers at the end of the collision detection process.
+
+An example of how this might be used is to add a local variable in your `PositionComponent` to save
+the other components with which this one is colliding:
```
```suggestion
the other components with which it's colliding with:
```

spydon commented on doc/flame/collision_detection.md:
````diff
@@ -120,6 +120,41 @@ other way around. If there are no intersections with the edges on a solid hitbox
+
+```
+
+Finally, one adds a listener to the `onLoad` method of the `PositionComponent` to call a routine
````
```suggestion
Finally, one adds a listener to the `onLoad` method of the `PositionComponent` to call a function
```

spydon commented on doc/flame/collision_detection.md:
````diff
@@ -120,6 +120,41 @@ other way around. If there are no intersections with the edges on a solid hitbox
+which will resolve how the collisions should be dealt with:
+
+```dart
+(game as HasCollisionDetection).collisionDetection.collisionsCompletedNotifier.
+                                addListener(() {
+  resolveCollisions();
+});
````
Formatting

christian-mindfulness replied:
I have to break the line (as it's too long otherwise).  I've now done it in a way that looks nice to me, happy to hear your thoughts.

spydon replied:
if you put it in an IDE and or https://dartpad.dev it will correctly format it for you, since it follows 80 width by default

christian-mindfulness replied:
Couldn't get VScode, dartpad or tutorialspoint to do it automatically for me. :-(

spydon commented on packages/flame/lib/src/collisions/collision_detection.dart:
```diff
@@ -161,3 +166,9 @@ abstract class CollisionDetection<T extends Hitbox<T>,
   });
 }
+
+/// Add a basic class which implements a Listener.
```
```suggestion
/// A class to handle callbacks for when the collision detection is done each tick.
```

spydon commented on packages/flame/test/collisions/collision_detection_test.dart:
```diff
@@ -1756,6 +1756,20 @@ void main() {
+        var calledTimes = 0;
+        final listeners = List.generate(10, (_) => () => calledTimes++);
+        for(final listener in listeners) {
+          game.collisionDetection.collisionsCompletedNotifier.addListener(listener);
```
Run `melos format` and it will fix this line for you.

spydon reviewed (commented):
A few final comments

christian-mindfulness commented:
Thanks @spydon for the review.  I think my latest commit addresses all the issues.  Let me know if I've missed anything.  I've also simplified the extension of `ChangeNotifier` to have an empty body, which avoids any warnings.

spydon commented:
> Thanks @spydon for the review. I think my latest commit addresses all the issues. Let me know if I've missed anything. I've also simplified the extension of `ChangeNotifier` to have an empty body, which avoids any warnings.

You have to override `notifyListeners` since `super.notifyListeners` isn't exposed, only for tests.

christian-mindfulness commented:
> > Thanks @spydon for the review. I think my latest commit addresses all the issues. Let me know if I've missed anything. I've also simplified the extension of `ChangeNotifier` to have an empty body, which avoids any warnings.
> 
> You have to override `notifyListeners` since `super.notifyListeners` isn't exposed, only for tests.

Curious that it worked for me - I've changed back.

spydon commented:
> Curious that it worked for me - I've changed back.

It works, but you get an analyzer warning. That was why the pipeline didn't pass.

spydon commented on doc/flame/collision_detection.md:
````diff
@@ -120,6 +120,42 @@ other way around. If there are no intersections with the edges on a solid hitbox
+which will resolve how the collisions should be dealt with:
+
+```dart
+(game as HasCollisionDetection).collisionDetection
+                               .collisionsCompletedNotifier
+                               .addListener(() {
+  resolveCollisions();
+});
````
```suggestion
(game as HasCollisionDetection)
    .collisionDetection
    .collisionsCompletedNotifier
    .addListener(() {
  resolveCollisions();
});
```
This is the correct formatting

spydon commented on packages/flame/lib/src/collisions/collision_detection.dart:
```diff
@@ -161,3 +166,10 @@ abstract class CollisionDetection<T extends Hitbox<T>,
+/// tick.
+class CollisionDetectionCompletionNotifier extends ChangeNotifier {
+  @override
+  void notifyListeners() => super.notifyListeners();
```
Did you not need any lint ignore here now? :thinking:

spydon replied:
The pipeline passes so I guess it is okay, but I wonder why it complained before, but not now...
